pkg/kubectl/cmd/util/clientcache.go

```diff
@@ -46,14 +49,41 @@ func (c *clientCache) ClientConfigForVersion(version string) (*client.Config, er
 	}
 	// TODO: have a better config copy method
 	config := *c.defaultConfig
-	if len(version) != 0 {
-		config.Version = version
+	version, err := c.negotiateVersion(&config, version)
+	if err != nil {
+		return nil, err
 	}
+	config.Version = version
 	client.SetKubernetesDefaults(&config)
 
 	return &config, nil
 }
 
+func (c *clientCache) negotiateVersion(config *client.Config, version string) (string, error) {
+	client, err := client.New(config)
+	if err != nil {
+		return "", err
+	}
+	serverVersions, err := client.ServerAPIVersions()
+	if err != nil {
+		return "", fmt.Errorf("couldn't read version from server: %v\n", err)
+	}
+	for _, serverVersion := range serverVersions.Versions {
+		if version == serverVersion {
+			return version, nil
+		}
+	}
+	for _, clientVersion := range registered.RegisteredVersions {
```

Comment:
If we hard fail when `len(version) != 0`, then a 0.19.0 kubectl binary will hard fail every command run against a server running < 0.19.0. That is not desirable behavior. It prints a warning that the client-requested api version is unsupported by the server and falls back to first available api version in order of preference (by client's registered versions).

Also added check for requested version being registered.